activesupport/lib/active_support/cache.rb

```diff
@@ -438,18 +438,15 @@ def fetch_multi(*names)
         options = merged_options(options)
 
         instrument :read_multi, names, options do |payload|
-          read_multi_entries(names, options).tap do |results|
-            payload[:hits] = results.keys
-            payload[:super_operation] = :fetch_multi
+          reads  = read_multi_entries(names, options)
+          writes = (names - reads.keys).map { |name| [name, yield(name)] }.to_h
 
-            writes = {}
+          payload[:hits] = reads.keys
+          payload[:super_operation] = :fetch_multi
 
-            (names - results.keys).each do |name|
-              results[name] = writes[name] = yield(name)
-            end
+          write_multi(writes, options)
 
-            write_multi writes, options
-          end
+          names.map { |name| [name, reads.fetch(name) { writes[name] }] }.to_h
```

Comment:
Tried grokking this over multiple attempts, but I find the two maps quite confusing since we seem to do similar yet different processing.

What if we kept some of the old flow ala this?

```ruby
instrument :read_multi, names, options do |payload|
  reads = read_multi_entries(names, options)
  payload[:hits] = reads.keys
  payload[:super_operation] = :fetch_multi

  writes  = {}
  ordered = names.each_with_object({}) do |name, hash|
    hash[name] = reads.fetch(name) { writes[name] = yield name }
  end

  write_multi writes, options
  ordered
end
```

(Probably ends up needing to be `yield(name)`. Psst, also a nice case for `index_with`, but we can't use core extensions here.)